Source/FieldSolver/ImplicitSolvers/WarpXImplicitOps.cpp

```diff
@@ -0,0 +1,333 @@
+        amrex::ParticleReal* uy_n = pti.GetAttribs(particle_comps["uy_n"]).dataPtr();
+        amrex::ParticleReal* uz_n = pti.GetAttribs(particle_comps["uz_n"]).dataPtr();
+
+        const long np = pti.numParticles();
+
+        amrex::ParallelFor( np, [=] AMREX_GPU_DEVICE (long ip)
+        {
+            amrex::ParticleReal xp, yp, zp;
+            getPosition(ip, xp, yp, zp);
+
+#if (AMREX_SPACEDIM >= 2)
+            x_n[ip] = xp;
+#endif
+#if defined(WARPX_DIM_3D) || defined(WARPX_DIM_RZ)
+            y_n[ip] = yp;
+#endif
+            z_n[ip] = zp;
+
+            ux_n[ip] = ux[ip];
+            uy_n[ip] = uy[ip];
+            uz_n[ip] = uz[ip];
+
+        });
+
+    }
+    }
+
+    }
+
+}
+
+void
+WarpX::FinishImplicitParticleUpdate ()
+{
+    using namespace amrex::literals;
+
+    const int lev = 0;
+
+    for (auto const& pc : *mypc) {
+
+#ifdef AMREX_USE_OMP
+#pragma omp parallel
+#endif
+    {
+
+    auto particle_comps = pc->getParticleComps();
+
+    for (WarpXParIter pti(*pc, lev); pti.isValid(); ++pti) {
+
+        const auto getPosition = GetParticlePosition(pti);
+        const auto setPosition = SetParticlePosition(pti);
+
+        auto& attribs = pti.GetAttribs();
+        amrex::ParticleReal* const AMREX_RESTRICT ux = attribs[PIdx::ux].dataPtr();
+        amrex::ParticleReal* const AMREX_RESTRICT uy = attribs[PIdx::uy].dataPtr();
+        amrex::ParticleReal* const AMREX_RESTRICT uz = attribs[PIdx::uz].dataPtr();
+
+#if (AMREX_SPACEDIM >= 2)
+        amrex::ParticleReal* x_n = pti.GetAttribs(particle_comps["x_n"]).dataPtr();
+#endif
+#if defined(WARPX_DIM_3D) || defined(WARPX_DIM_RZ)
+        amrex::ParticleReal* y_n = pti.GetAttribs(particle_comps["y_n"]).dataPtr();
+#endif
+        amrex::ParticleReal* z_n = pti.GetAttribs(particle_comps["z_n"]).dataPtr();
+        amrex::ParticleReal* ux_n = pti.GetAttribs(particle_comps["ux_n"]).dataPtr();
+        amrex::ParticleReal* uy_n = pti.GetAttribs(particle_comps["uy_n"]).dataPtr();
+        amrex::ParticleReal* uz_n = pti.GetAttribs(particle_comps["uz_n"]).dataPtr();
+
+        const long np = pti.numParticles();
+
+        amrex::ParallelFor( np, [=] AMREX_GPU_DEVICE (long ip)
+        {
+            amrex::ParticleReal xp, yp, zp;
+            getPosition(ip, xp, yp, zp);
+
+#if (AMREX_SPACEDIM >= 2)
+            xp = 2._rt*xp - x_n[ip];
+#endif
+#if defined(WARPX_DIM_3D) || defined(WARPX_DIM_RZ)
+            yp = 2._rt*yp - y_n[ip];
+#endif
+            zp = 2._rt*zp - z_n[ip];
+
+            ux[ip] = 2._rt*ux[ip] - ux_n[ip];
+            uy[ip] = 2._rt*uy[ip] - uy_n[ip];
+            uz[ip] = 2._rt*uz[ip] - uz_n[ip];
+
+            setPosition(ip, xp, yp, zp);
+        });
+
+    }
+    }
+
+    }
+
+}
+
+void
+WarpX::FinishImplicitField( amrex::Vector<std::array< std::unique_ptr<amrex::MultiFab>, 3 > >& Field_fp,
+                      const amrex::Vector<std::array< std::unique_ptr<amrex::MultiFab>, 3 > >& Field_n,
+                      amrex::Real theta )
+{
```

Comment:
Could you add a comment here, explaining what this function does
(i.e. at the end of a nonlinear solve, we get the field at n+theta. We then need to compute the fields n+1, which is what this function does.)

Reply:
Done